### PR TITLE
Support parallel execution of time-chunked queries

### PR DESCRIPTION
**Pull request by steve-scalyr**

### CODE REVIEW

File: src/main/com/scalyr/api/query/QueryService.java [hunk 1]
```diff
@@ -74,13 +69,34 @@ public QueryService(String apiToken) {
    *     A reasonable chunk size is often 12 to 24 hours. Smaller values reduce the possibility of timeout, but add
    *     overhead for extra round-trips to the Scalyr server.
    *
-   *    chunkSizeHours is only supported for a limit set of query types; others will thrown a RuntimeException.
+   *     chunkSizeHours is only supported for a limit set of query types; others will throw a RuntimeException.
+   * @param maxParallelThreads Only applicable when chunkSizeHours is used. If greater than 1, then we will run this
+   *     many queries in parallel. This should only be used when querying extremely large datasets; generally consult
+   *     with Scalyr support before specifying a nonzero value for this parameter. In any case, this parameter should
```
Comment (goldfrapp04) [inline src/main/com/scalyr/api/query/QueryService.java:75]: `s/nonzero/greater-than-1/`?

File: src/main/com/scalyr/api/query/QueryService.java [hunk 2]
```diff
@@ -128,20 +144,49 @@ public LogQueryResult logQuery(String filter, String startTime, String endTime,
 
     LogQueryResult merged = null;
     List<Pair<String>> chunkList = chunked.collect(Collectors.toList());
-    for (int i=0; i<chunkList.size(); i++) {
+
+    if (queryThreadpool != null) {
+      return parallelLogQuery(filter, maxCount, pageMode, columns, continuationToken, chunkList);
+    } else {
```
Comment (goldfrapp04) [inline src/main/com/scalyr/api/query/QueryService.java:150]: Remove the `else {}` wrapping around the following code block? We've already early-returned if parallel-querying anyways.

File: src/main/com/scalyr/api/query/QueryService.java [hunk 3]
```diff
@@ -128,20 +144,49 @@ public LogQueryResult logQuery(String filter, String startTime, String endTime,
...
+  }
+
+  /**
+   * Query the time ranges specified by chunkList, and return the merged result. Individual queries will be executed in
+   * parallel, using queryThreadpool.
+   */
+  private LogQueryResult parallelLogQuery(String filter, Integer maxCount, PageMode pageMode, String columns, String continuationToken, List<Pair<String>> chunkList) {
+    List<Future<LogQueryResult>> futures = new ArrayList<>();
+    AtomicBoolean done = new AtomicBoolean(false);
+    for (int i = 0; i < chunkList.size(); i++) {
```
Comment (goldfrapp04) [inline src/main/com/scalyr/api/query/QueryService.java:168]: Can just be `chunkList.forEach(pair -> ...)`

File: src/main/com/scalyr/api/query/QueryService.java [hunk 4]
```diff
@@ -128,20 +144,49 @@ public LogQueryResult logQuery(String filter, String startTime, String endTime,
 
     LogQueryResult merged = null;
     List<Pair<String>> chunkList = chunked.collect(Collectors.toList());
-    for (int i=0; i<chunkList.size(); i++) {
+
+    if (queryThreadpool != null) {
+      return parallelLogQuery(filter, maxCount, pageMode, columns, continuationToken, chunkList);
+    } else {
+      for (int i = 0; i < chunkList.size(); i++) {
```
Comment (goldfrapp04) [inline src/main/com/scalyr/api/query/QueryService.java:151]: Can just be `for (Pair<String> pair : chunkList) ...`

**Review by goldfrapp04 — COMMENTED** [review]
Looks good, all my comments are cosmetics.

File: src/main/com/scalyr/api/query/QueryService.java (hunk 1 above)
Comment (steve-scalyr) [reply src/main/com/scalyr/api/query/QueryService.java:75]: Fixed

File: src/main/com/scalyr/api/query/QueryService.java (hunk 2 above)
Comment (steve-scalyr) [reply src/main/com/scalyr/api/query/QueryService.java:150]: Fixed

File: src/main/com/scalyr/api/query/QueryService.java (hunk 3 above)
Comment (steve-scalyr) [reply src/main/com/scalyr/api/query/QueryService.java:168]: ...right. Fixed.

File: src/main/com/scalyr/api/query/QueryService.java (hunk 4 above)
Comment (steve-scalyr) [reply src/main/com/scalyr/api/query/QueryService.java:151]: Fixed

**Comment by steve-scalyr** [comment]
@goldfrapp04 Thanks for the quick turnaround. Ready for another look (and I'm done with the changes I had planned).

**Review by goldfrapp04 — APPROVED** [review]

**Pull request merged**